### ISSUE TITLE
Rspec error VCR::Errors::UnhandledHTTPRequestError

### ISSUE THREAD

**Issue by StephaneJuban** [paulsingh/angellist-api#35]
I get these two errors when I run the ```bundle exec rake spec``` command :

```
Failure/Error: results = client.search('joshua', :type => 'User')
VCR::Errors::UnhandledHTTPRequestError

Failure/Error: roles = client.get_startup_roles(params)
VCR::Errors::UnhandledHTTPRequestError
```

And same with the new feature I just added. Is it my config or something else ?

**Comment by ches** [comment]
Hi Stéphane, that's strange, I can't immediately reproduce that but I've got a pretty old `Gemfile.lock` locally and I'd be slightly suspicious that a newer version of VCR satisfying the `~> 2.1` dependency might be matching the requests differently for some reason, that's all I can think of.

What's the actual version of VCR that's been installed in your dev environment according to `Gemfile.lock`?

About your new additions, has VCR recorded the HTTP transaction? It should record on the first test run with the `:record => :once` setting in `spec/support/vcr.rb`. If so then hopefully getting to the bottom of the failure on existing tests will make the answer clear...

Also, does your new feature require authentication? That's an awkward aspect of the integration test setup but I haven't really thought of a better way to do it, open to ideas.

**Comment by StephaneJuban** [comment]
Hi ches,

Thanks for digging into it. FYI I use Cloud9IDE and in my Gemfile.lock I got ```vcr (2.9.3)```
How do I know that VCR record the HHTP transaction ?
My new feature does not require authentication, in fact this is the [Users/Tags](https://angel.co/api/spec/users#GET_tags_%3Aid_users) feature.

Here is [the complete error log](http://pastebin.com/JbLT67A0). The first two error are here from the beginning and the third is added by me but they are all the same.

**Comment by ches** [comment]
> How do I know that VCR record the HHTP transaction ?

Basically just that there should be a changed or new untracked file in `spec/fixtures/cassettes`, for the new transaction.

Thanks for the full error trace! [This recording][1] ought to be matching for the first failing case, but the query parameters are switched. My hunch, from seeing both cases, is that maybe VCR started sorting query params alphabetically to be deterministic, or something like that.

When I have a chance I'll update VCR in my bundle and try this. If that turns out to be it then I'll update the VCR cassettes and push a gemspec with a stricter version requirement for VCR. (My lock version current is 2.4.0 FWIW). Thanks for your patience!

[1]: https://github.com/paulsingh/angellist-api/blob/405243da95e9d314fb1a22fc9588fad8290f6383/spec/fixtures/cassettes/search.yml#L112

**Comment by StephaneJuban** [comment]
> Basically just that there should be a changed or new untracked file in spec/fixtures/cassettes, for the new transaction.

Just so you know, the answer is no, ```spec/fixtures/cassettes/users.yml``` is not modified for me.

**Comment by ches** [comment]
Sorry for the delay Stéphane, I reproduced this after nuking my `Gemfile.lock` entirely, and the hunch about sorting params was on the mark, but [Faraday was the culprit][1], not VCR. I believe the change referenced above should fix the issues for you in the tests, please report back if not!

> Just so you know, the answer is no, spec/fixtures/cassettes/users.yml is not modified for me.

So this workflow is a little awkward, but it's just the way it goes for sanity with VCR: we use [the `:once` record mode][2], so if you're adding new tests to an existing file (and thus using an existing cassette), you'll need to do one of two things first:

1. Temporarily update the VCR tag on the `describe` block to change the mode (e.g. `:vcr => { cassette_name => 'users', :record => :new_episodes }`), run the test to record, then reset the option before committing.
2. Just nuke the cassette file entirely with `rm` and then run the test to regenerate it from scratch, so `:once` works fine. Commit the entire file change. I usually go this route since it's easier and it's good to refresh the recordings of old requests from time to time anyway.

Hope that makes sense :smile: 

[1]: https://github.com/lostisland/faraday/issues/353
[2]: https://www.relishapp.com/vcr/vcr/v/2-9-3/docs/record-modes/once

**Comment by ches** [comment]
Err, a little update to that workflow, as if it wasn't confusing enough: even if you delete an entire cassette file, it seems that the `:once` mode is now only recording for a single test within a `describe` block and then throwing errors on every spec after the first :weary: 

I'm pretty certain that `:once` used to apply for an entire `describe` block, possibly this has changed along the way in VCR or perhaps there have been changes to RSpec's handling of metadata for `describe` in newer versions that changed the semantics, I'm not sure.

So, in short, you'll need to temporarily add the `:record => :new_episodes` either way (I'd still recommend deleting the cassette first to completely refresh it). I'll give some thought to making `:new_episodes` the default, but my experience with that in the past has been that people accidentally record new, incorrect requests while getting newly added tests to work, and that often results in dubious/unreliable tests...

Sorry for the confusion, I'll add something about this to the README for contributors once I work out what the preferable and hopefully straightforward procedure is.

**Comment by StephaneJuban** [comment]
Hi ches,

Thank you for your help and the update! However I'm still not able to handle the tests process properly sorry :(

First I update my branch to v1.1.0. With your changes all initials tests are OK. That's a good start.
But if I did what you suggest, deleting the entire cassette file user.yml and add `:vcr => { :cassette_name => 'users', :record => :new_episodes }`.

First, this file is not regenerated, and secondly all tests (users_spec.rb) failed and the new one I added is ignored.

What is your thought about that, am I missing something obvious? Sorry, I'm not familiar with tests...
